bentoml/artifact/onnx_model_artifact.py

```diff
@@ -0,0 +1,174 @@
+
+    >>> # Convert into ONNX format
+    >>> from skl2onnx import convert_sklearn
+    >>> from skl2onnx.common.data_types import FloatTensorType
+    >>> initial_type = [('float_input', FloatTensorType([None, 4]))]
+    >>>
+    >>> onnx_model = convert_sklearn(clr, initial_types=initial_type)
+    >>> with open("rf_iris.onnx", "wb") as f:
+    >>>     f.write(onnx_model.SerializeToString())
+    >>>
+    >>>
+    >>> import numpy
+    >>> import bentoml
+    >>> from bentoml.artifact import OnnxModelArtifact
+    >>> from bentoml.handlers import DataframeHandler
+    >>>
+    >>> @bentoml.env(auto_pip_dependencies=True)
+    >>> @bentoml.artifact([OnnxModelArtifact('model')])
+    >>> class OnnxIrisClassifierService(bentoml.BentoService):
+    >>>     @bentoml.api(DataframeHandler)
+    >>>     def predict(self, df):
+    >>>         input_data = df.to_numpy().astype(numpy.float32
+    >>>         input_name = self.artifacts.model.get_inputs()[0].name
+    >>>         output_name = self.artifacts.model.get_outputs()[0].name
+    >>>         return self.artifacts.model.run(
+    >>>                     [output_name], {input_name: input_data}
+    >>>                )[0]
+    >>>
+    >>> svc = OnnxIrisClassifierService()
+    >>>
+    >>> # Option one: pack with path to model on local system
+    >>> svc.pack('model', './rf_iris.onnx')
+    >>>
+    >>> # Option two: pack with ONNX model object
+    >>> # svc.pack('model', onnx_model)
+    >>>
+    >>> # Save BentoService
+    >>> svc.save()
+    """
+
+    def __init__(self, name, backend='onnx'):
+        super(OnnxModelArtifact, self).__init__(name)
+        self.backend = backend
+
+    def _model_file_path(self, base_path):
+        return os.path.join(base_path, self.name + '.onnx')
+
+    def pack(self, obj):  # pylint:disable=arguments-differ
+        try:
+            import onnx
+        except ImportError:
+            raise MissingDependencyException(
+                'onnx package is required for packing OnnxModelArtifact'
+            )
+        if isinstance(obj, onnx.ModelProto):
+            return _OnnxModelArtifactWrapper(self, obj)
+        else:
+            return _ExportedOnnxModelArtifact(self, obj)
+
+    def load(self, path):
+        if self.backend == 'onnx':
+            inference_model = _load_onnx_saved_model(self._model_file_path(path))
+        else:
+            raise NotImplementedError(
+                f'{self.backend} as onnx runtime is not supported at the moment'
+            )
+        return self.pack(inference_model)
+
+    @property
+    def pip_dependencies(self):
+        dependencies = ['onnx']
+        if self.backend == 'onnx':
+            dependencies.append('onnxruntime')
+        return dependencies
+
+
+class _OnnxModelArtifactWrapper(BentoServiceArtifactWrapper):
+    def __init__(self, spec, model):
+        super(_OnnxModelArtifactWrapper, self).__init__(spec)
+        self._model = model
+
+    def get(self):
```

Comment:
returning a `ModelProto` here is really bad because it has a completely different signature than a loaded model(e.g. loaded from onnx runtime's inferencesession for example)

Might be better to initialize a backend from here with the `ModelProto`, you can do it lazily similar to the `_ExportedOnnxModelArtifact `

https://github.com/microsoft/onnxruntime/pull/718

Reply:
I think this is supported in current onnxruntime release now: https://github.com/microsoft/onnxruntime/issues/441#event-2334251059

Reply:
```
if not self._model:
      if self.backend == 'onnxruntime':
            self._model = onnxruntime.InferenceSession(self._model_proto)
      else:
            raise...
     
return self._model
```

Reply:
I haven't tested this code though, could you try it out?